### PR TITLE
Docker image: Multi arch support

### LINKED ISSUE #4599: Add aarch64 docker image
The official docker image matrixdotorg/synapse only works with x86_64. I successfully started the docker file on my aarch64 machine by simply changing the images from `docker.io/python` to `docker.io/arm64v8/python` in the Dockerfile.

Please provide this prebuilt as well :)

### PR DESCRIPTION
**Pull request by Starbix**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

This PR updates the CircleCI configuration to use `docker buildx` to build the docker image. This way multi-arch images can be built on amd64. Currently only amd64 and arm64v8 are enabled. But other architectures such as arm/v7 or even s390x can be easily added.

Building for amd64 and arm64v8 took around 1h 30m on a Docker Medium instance on CircleCI.

Closes #4599

In case you want to test it, the multiarch image created by CircleCI is available under `starbix/synapse:latest` (master branch).

Signed-off-by: Cédric Laubacher <cedric@laubacher.io>

### CODE REVIEW

**Comment by Starbix** [comment]
I've added CircleCI commands to remove duplicate code

**Comment by babolivier** [comment]
This looks globally sane to me, however when I try to run it in CircleCI it seems to be stuck on the `pip install`s. I'll try to find some time to investigate why.

**Comment by babolivier** [comment]
Also you can fix the CI by pulling and merging the develop branch into your branch.

**Comment by Starbix** [comment]
> This looks globally sane to me, however when I try to run it in CircleCI it seems to be stuck on the `pip install`s. I'll try to find some time to investigate why.

How long did you let it run for? It may just be slow

**Comment by babolivier** [comment]
It's currently been running for 14min.

**Comment by babolivier** [comment]
Oh wait it is indeed very slow, the amd64 passed that step, now it's just waiting for the arm64 one. I'll let it run and see what happens.

**Comment by Starbix** [comment]
It can take 1.5 hours. Maybe only enabling multi-arch for releases would make sense.

**Comment by anoadragon453** [comment]
We've decided this is something we want, but are aware that it would delay docker images after a release is made. We need to see whether this can be sped up at all, but if not we'll probably still do it.

**Comment by Starbix** [comment]
Using a native arm64 machine as a [self-hosted runner](https://help.github.com/en/actions/hosting-your-own-runners/adding-self-hosted-runners) would speed things up.

File: .circleci/config.yml [hunk 1]
```diff
@@ -31,3 +37,40 @@ workflows:
...
+      buildx_version:
+        type: string
+        default: "v0.4.1"
+    steps:
+      - run: apk add --no-cache curl
+      - run: mkdir -vp ~/.docker/cli-plugins/ ~/dockercache
+      - run: curl --silent -L "https://github.com/docker/buildx/releases/download/<< parameters.buildx_version >>/buildx-<< parameters.buildx_version >>.linux-amd64" > ~/.docker/cli-plugins/docker-buildx
+      - run: chmod a+x ~/.docker/cli-plugins/docker-buildx
+      - run: docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
+      - run: docker context create old-style
```
Comment (richvdh) [inline .circleci/config.yml:54]: why is the context called `old-style` ? (indeed, why do we need to create a context here?)

File: .circleci/config.yml [hunk 2]
```diff
@@ -31,3 +37,40 @@ workflows:
...
+
+  docker_build:
+    description: Builds and pushed images to dockerhub using buildx
+    parameters:
+      dockerfile:
+        type: string
+        default: docker/Dockerfile
+      push:
+        type: boolean
+        default: true
```
Comment (richvdh) [inline .circleci/config.yml:65]: is there any point in parameterising these? it seems to complicate the commandline without any real benefit.

File: .circleci/config.yml [hunk 3]
```diff
@@ -31,3 +37,40 @@ workflows:
...
+        type: string
+        default: linux/arm64/v8,linux/amd64
+      context:
+        type: string
+        default: .
+      parameters:
+        type: string
+        default: --label gitsha1=${CIRCLE_SHA1}
+    steps:
+      - run: docker buildx build -f << parameters.dockerfile >><<# parameters.push >> --push<</ parameters.push >> --platform << parameters.platforms >> << parameters.parameters >> << parameters.context >>
```
Comment (richvdh) [inline .circleci/config.yml:76]: the default output format does not work well in circleci. suggest `--progress=plain`.

**Comment by richvdh** [comment]
(link to test circleci build: https://app.circleci.com/pipelines/github/matrix-org/synapse/13909/workflows/b387ce2c-5230-4505-8bea-e8dad635adad/jobs/25549)

**Comment by richvdh** [comment]
Now that #7839 has landed, I've kicked off another test build at https://app.circleci.com/pipelines/github/matrix-org/synapse/14292/workflows/0c35f3e6-05fe-42e3-a955-30712a8c13fa/jobs/25558. I'll be interested to see how long it takes this time.

File: .circleci/config.yml [hunk 4]
```diff
@@ -1,24 +1,30 @@
...
 jobs:
   dockerhubuploadrelease:
-    machine: true
+    docker:
+      - image: docker:git
     steps:
       - checkout
-      - run: docker build -f docker/Dockerfile --label gitsha1=${CIRCLE_SHA1} -t matrixdotorg/synapse:${CIRCLE_TAG} -t matrixdotorg/synapse:${CIRCLE_TAG}-py3 .
+      - setup_remote_docker:
+          version: 18.09.3
```
[the remark this answers is not in the document]
Comment (Starbix) [reply .circleci/config.yml:9]: Sorry for not responding earlier, I wasn't near a computer for the last 2 weeks.
If I remember correctly, 18.09 is the highest version CircleCI supports and it defaulted to a lower version.
A higher version would be beneficial, as there wouldn't be a need to create context.

File: .circleci/config.yml (hunk 1 above)
Comment (Starbix) [reply .circleci/config.yml:54]: The name could be anything, but it was often called `old-style` in the tutorials I read. Creating a context is needed as CircleCI isn't running a relatively modern version of Docker.

File: .circleci/config.yml (hunk 2 above)
Comment (Starbix) [reply .circleci/config.yml:65]: No, not at this moment. I added these parameters so one could easily expand the pipeline. I can remove the unused parameters.

File: .circleci/config.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Starbix) [reply .circleci/config.yml:9]: The newest version available seems to be 19.03 now: https://circleci.com/docs/2.0/building-docker-images/#docker-version
I'll be able to test it and make the other changes this week.

**Comment by richvdh** [comment]
> I'll be interested to see how long it takes this time.

answer: it failed, for missing header files. Looks like more changes are needed to support arm64 builds here.

**Comment by richvdh** [comment]
I think this is superceded by #7921, so I'm going to close it.

**Pull request closed without merging**